src/services/connection.service.ts

```diff
@@ -0,0 +1,221 @@
+        for (const guid of this.routingTable.peers.keys()) {
+            promises.push(this.sendMessage(message, guid, responseImplementation));
+        }
+        return Promise.all(promises);
+    }
+
+    /**
+     * Check of messageType of the given message has a known implementation, and executes the implementation.
+     *
+     * @param message - The incoming message
+     * @param senderGuid - The GUID of the sender
+     * @param senderIp - The IP of the sender
+     */
+    public onMessage(message: Message): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            if (!this.sender) {
+                reject();
+                return;
+            }
+            logger.info(`${message.type} message received`);
+            
+            const responseImplementation = this.requestsMap.get(message.correlationId);
+            if (responseImplementation) {
+                await responseImplementation(message);
+                const responseDeffered = this.responseDefferedsMap.get(message.correlationId);
+                if (responseDeffered) {
+                    responseDeffered.resolve!(true);
+                }
+            }
+
+            const implementation = this.receiveHandlers.get(message.type);
+            if (implementation && typeof implementation === "function") {
+                await implementation(message, message.originalSenderGuid, (responseMessage: Message) => {
+                    responseMessage.correlationId = message.guid;
+                    responseMessage.originalSenderGuid = message.originalSenderGuid;
+                    this.sendMessage(responseMessage, responseMessage.originalSenderGuid);
+                });
+                // Acknowledge this message
+                if (message.type !== MessageType.ACKNOWLEDGE) {
+                    const destination = this.getIpFromRoutingTable(message.originalSenderGuid);
+                    this.sender.sendAcknowledgeMessage(message, destination);
+                }
+            }
+            resolve();
+        });
+    }
+
+    /**
+     * Leave the network.
+     */
+    public leave(guid: string): void {
+        const message = new Message(MessageType.LEAVE, guid);
+        this.sendBroadcast(message);
+    }
+
+
+    public sendMessageByIp(message: Message, destinationIp: string, responseImplementation?: RESPONSE_TYPE): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            if (!this.sender) {
+                reject();
+                return;
+            }
+            
+            if (responseImplementation) {
+                this.requestsMap.set(message.guid, responseImplementation);
+                this.responseDefferedsMap.set(message.guid, new Deferred());
+            }
+            await this.sender.sendMessage(message, destinationIp);
+            resolve();
+        });
+    }
+
+    public getPromiseForResponse(message: Message): Promise<void> {
```

Comment:
Please add an explanation in the docs as to why we're using this function and why we're wrapping the deferreds in an extra promise.